libhb/rendersub.c

```diff
@@ -115,6 +131,112 @@ hb_filter_object_t hb_filter_render_sub =
     .close         = hb_rendersub_close,
 };
 
+static void hb_rect_vec_resize(hb_rect_vec_t *list, int size)
+{
+    hb_rect_t *rects = realloc(list->rects, size * sizeof(hb_rect_t));
+    if (rects == NULL)
+    {
+        return; // Error.  Should never happen.
+    }
+    list->rects = rects;
+    list->size = size;
+}
+
+static void hb_rect_vec_merge_overlapping_rects(hb_rect_vec_t *list)
+{
+    if (list->count < 2)
+    {
+        return;
+    }
+
+    // Merge nearby rects to avoid
+    // having too many around
+    const int offset = 8;
+    hb_rect_t empty = {0};
+
+    for (int i = 0; i < list->count - 1; i++)
+    {
+        hb_rect_t a = list->rects[i];
+        for (int j = i + 1; j < list->count; j++)
+        {
+            hb_rect_t b = list->rects[j];
+            if (a.x - offset <= b.x + b.width  && a.x + a.width  + offset >= b.x &&
+                a.y - offset <= b.y + b.height && a.y + a.height + offset >= b.y)
+            {
+                int width  = FFMAX(a.width + a.x, b.width + b.x);
+                int height = FFMAX(a.height + a.y, b.height + b.y);
+                a.x      = FFMIN(a.x, b.x);
+                a.y      = FFMIN(a.y, b.y);
+                a.width  = width  - a.x;
+                a.height = height - a.y;
```

Comment:
I think you can simplify this block of logic by using solely coordinates.

```c
typedef struct hb_box_s {
int x1, y1, x2, y2;
} hb_box_t;

static inline bool do_intersect(const hb_box_t *a, const hb_box_t *b, int offset)
{
    return ((FFMIN(a->x2, b->x2) + offset - FFMAX(a->x1, b->x1)) > 0) && ((FFMIN(a->y2, b->y2) + offset - FFMAX(a->y1, b->y1)) > 0);
}

static inline void union(hb_box_t *a, const hb_box_t *b)
{
    if (a->x1 > b->x1) { a->x1 = b->x1; }
    if (a->y1 > b->y1) { a->y1 = b->y1; }
    if (a->x2 < b->x2) { a->x2 = b->x2; }
    if (a->y2 < b->y2) { a->y2 = b->y2; }
}

// ....

if (do_intersect(a, b, 8))
{
    union(a, b);
}
```

You only deal with the width and height when you create all of the small rects and when you render the definitive ones.

Reply:
Thanks, I have integrated your suggestions.